### PR TITLE
api: Implement state diff

### PR DESCRIPTION
**Pull request by Desahri**
Includes a test

### CODE REVIEW

**Comment by helaan** [comment]
Hmm, this PR is a bit limited: you could also just send the properties that changed, which would save even more bandwidth.
Example: If the `batteryStatus` property changes of turtle 1, you can just send `{"turtles"{"1":{"batteryStatus":99}}}`

File: pkg/api/api.go [hunk 1]
```diff
@@ -219,3 +219,38 @@ func NewMessage(typ MessageType, pld json.RawMessage, parentID *ulid.ULID) *Mess
...
+	r := make(map[string]*TurtleState)
+	for k, v := range turtlesNew {
+		if !turtleStateIsSame(v, turtlesOld[k]) {
+			r[k] = v
+		}
+	}
+	return r
+}
+
+func turtleStateIsSame(ts1, ts2 *TurtleState) bool {
```
Comment (rvolosatovs) [inline pkg/api/api.go]: This function is not needed, you can simply use `reflect.DeepEqual`

File: pkg/api/api.go [hunk 2]
```diff
@@ -219,3 +219,38 @@ func NewMessage(typ MessageType, pld json.RawMessage, parentID *ulid.ULID) *Mess
 		Payload:   pld,
 	}
 }
+
+//returns a turtleState mapping which contains only the turtleState of the turtles of which the old state is different from the new state
```
Comment (rvolosatovs) [inline pkg/api/api.go]: `// StateDifference ...`

File: pkg/api/api.go [hunk 3]
```diff
@@ -219,3 +219,38 @@ func NewMessage(typ MessageType, pld json.RawMessage, parentID *ulid.ULID) *Mess
...
+func StateDifference(turtlesOld, turtlesNew map[string]*TurtleState) map[string]*TurtleState {
```
Comment (rvolosatovs) [inline pkg/api/api.go]: Call it `StateDiff`

**Review by rvolosatovs — CHANGES_REQUESTED** [review]

**Comment by rvolosatovs** [comment]
As we discussed with @Huubvr earlier, it should be done in a following way:
1. Marshal both states into `map[string]interface{}` recursively https://godoc.org/github.com/fatih/structs#Map
2. Compute diff of these maps(recursively)
3. Find a way to go back from map to the state struct. Maybe we can marshal the map to json and the unmarshal the json into the state.

File: pkg/api/api_test.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
+package api
+
+import (
+	"github.com/rvolosatovs/turtlitto/pkg/api/apitest"
```
Comment (golangci) [inline pkg/api/api_test.go]: File is not goimports-ed

File: pkg/api/api_test.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+			inputNewState: &State{
+				Command: CommandStop,
+				Turtles: map[string]*TurtleState{
+					"1": {
+						Kinect1State:    KinectStateNoBall,
+						EmergencyStatus: apitest.Uint8Ptr(0),
+					},
+				},
+			},
+			expectedOutputState: &State{},
```
Comment (rvolosatovs) [inline pkg/api/api_test.go]: This has to be `nil`

**Review by rvolosatovs — CHANGES_REQUESTED** [review]

File: pkg/api/api_test.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+		{
+			inputOldState: &State{
+				Command: CommandStop,
+				Turtles: map[string]*TurtleState{
+					"1": {
+						Kinect1State:    KinectStateNoBall,
+						EmergencyStatus: apitest.Uint8Ptr(0),
+					},
+					"2": {
+						Kinect1State:    KinectStateBall,
```
Comment (rvolosatovs) [inline pkg/api/api_test.go]: For this test case add some field, which is included in the `old` struct, but not defined in the `new`, to make sure it's not removed.

File: pkg/api/api.go [hunk 7]
```diff
@@ -219,3 +222,41 @@ func NewMessage(typ MessageType, pld json.RawMessage, parentID *ulid.ULID) *Mess
...
+	}
+	keys := make(map[string]struct{})
+	for k := range old {
+		keys[k] = struct{}{}
+	}
+	for k := range new {
+		keys[k] = struct{}{}
+	}
+	for k := range keys {
+		if reflect.DeepEqual(old[k], new[k]) { //if there is a difference in the turtle state
```
Comment (rvolosatovs) [inline pkg/api/api.go]: This comment is not required and does not make too much sense, given that this function does not know anything about turtle states

**Pull request closed without merging**